src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs

```diff
@@ -89,44 +97,38 @@ async ValueTask SynchronizeSolutionAssetsWorkerAsync()
+
+            // then grab the information we want off of the compilation state object.
+            var solutionStateChecksum = await this.GetAssetAsync<SolutionStateChecksums>(
+                AssetPathKind.SolutionStateChecksums, compilationStateChecksums.SolutionState, cancellationToken).ConfigureAwait(false);
+
+            // Then grab the data we need off of hte state checksums
+            using var _1 = ArrayBuilder<Task>.GetInstance(out var tasks);
+            tasks.Add(this.GetAssetAsync<SourceGeneratorExecutionVersionMap>(
+                AssetPathKind.SolutionSourceGeneratorExecutionVersionMap, compilationStateChecksums.SourceGeneratorExecutionVersionMap, cancellationToken).AsTask());
+            tasks.Add(this.GetAssetAsync<SolutionInfo.SolutionAttributes>(
+                AssetPathKind.SolutionAttributes, solutionStateChecksum.Attributes, cancellationToken).AsTask());
+            tasks.Add(this.GetAssetsAsync<AnalyzerReference>(
+                AssetPathKind.SolutionAnalyzerReferences, solutionStateChecksum.AnalyzerReferences, cancellationToken));
+            await Task.WhenAll(tasks).ConfigureAwait(false);
 
             // Note: this search will be optimized on the host side.  It will search through the solution level values,
             // and then the top level project-state-checksum values only.  No other project data or document data will be
             // looked at.
-            await this.SynchronizeAssetsAsync<object, Dictionary<Checksum, object>>(
-                assetPath: AssetPathKind.SolutionState | AssetPathKind.ProjectStateChecksums,
-                checksums,
-                static (checksum, asset, checksumToObjects) => checksumToObjects.Add(checksum, asset),
-                arg: checksumToObjects, cancellationToken).ConfigureAwait(false);
+
+            using var _2 = PooledDictionary<Checksum, ProjectStateChecksums>.GetInstance(out var checksumToProjectStateChecksums);
+            await this.GetAssetsAsync<ProjectStateChecksums, Dictionary<Checksum, ProjectStateChecksums>>(
+                AssetPathKind.ProjectStateChecksums,
+                solutionStateChecksum.Projects.Checksums,
+                static (checksum, asset, checksumToProjectStateChecksums) => checksumToProjectStateChecksums.Add(checksum, asset),
+                arg: checksumToProjectStateChecksums, cancellationToken).ConfigureAwait(false);
 
             using var _3 = ArrayBuilder<ProjectStateChecksums>.GetInstance(out var allProjectStateChecksums);
 
             // fourth, get all projects and documents in the solution 
-            foreach (var (projectChecksum, projectId) in stateChecksums.Projects)
-            {
-                var projectStateChecksums = (ProjectStateChecksums)checksumToObjects[projectChecksum];
-                Contract.ThrowIfTrue(projectStateChecksums.ProjectId != projectId);
+            foreach (var (_, projectStateChecksums) in checksumToProjectStateChecksums)
```

Comment:
>checksumToProjectStateChecksums [](http://example.com/codeflow?start=55&length=31)

Is checksumToProjectStateChecksums even needed (couldn't the ProjectStateChecksums be directly added to projectStateChecksums)?

Reply:
yes.  much nicer.